### ISSUE TITLE
List support in Saxy.Builder

### ISSUE THREAD

**Issue by jcelliott** [qcam/saxy#86]
It looks like `Saxy.Builder` doesn't handle Lists by default. I was able to get it to work with this (hacky) implementation:

```elixir
defimpl Saxy.Builder, for: List do
  def build(nil), do: ""
  def build([]), do: ""

  def build(list) do
    list
    |> Enum.map(fn elem -> elem |> Saxy.Builder.build() |> Saxy.encode!() end)
    |> Enum.join()
  end
end
```

I was wondering if there's a better way to do this, and if it should be included in Saxy by default.

**Comment by qcam** [comment]
I'm not entirely sure what you wanted to do. Did you try to build 2 XML documents at once?

The idea with `Saxy.Builder` is that it transforms Elixir terms into XML simple form, and from there performs XML encoding with `Saxy.encode!`. Lists weren't supported from the start since I couldn't think of any good use case of it. If there is I'm very much open for adding it though.

**Comment by jcelliott** [comment]
I'm using it with an Ecto `embedded_schema` (just a struct under the hood). For the most part it works perfectly with the embedded schema, but I was trying to serialize a nested schema that uses `embeds_many` to associate a list of nested objects. Here's an example:

```elixir
defmodule Section do
  use Ecto.Schema
  @derive {Saxy.Builder, name: "section", attributes: [:name], children: [:widgets]}

  embedded_schema do
    field :name, :string
    embeds_many :widgets, Widget, on_replace: :delete
  end
end

defmodule Widget do
  use Ecto.Schema
  @derive {Saxy.Builder, name: "widget", attributes: [:type], children: [:content]}

  embedded_schema do
    field :type, :string
    field :content, :string
  end
end
```

When serializing this struct:
```elixir
%Section{
  name: "testing",
  widgets: [
    %Widget{type: text, content: "foo bar"},
    %Widget{type: image, content: "baz.gif"},
  ]
}
```

I would expect something like this output:
```
<section name="testing">
  <widgets>
    <widget type="text">foo bar</widget>
    <widget type="image">baz.gif</widget>
  </widgets>
</section>
```

Any thoughts? Maybe this isn't a common need.

**Comment by qcam** [comment]
I think this is a legit use case that `Saxy.Builder` should support 👍. I have a draft to tackle this at #87.

For instance, to transform the output in your example, we could change the code a bit to:

```elixir
defmodule Section do
  use Ecto.Schema
  @derive {
    Saxy.Builder,
    name: "section",
    attributes: [:name],
    children: [widgets: &__MODULE__.build_widget_xml/1]
  }

  embedded_schema do
    field :name, :string
    embeds_many :widgets, Widget, on_replace: :delete
  end

  def build_widget_xml(widgets) do
    Saxy.XML.element("widgets", [], widgets)
  end
end

defmodule Widget do
  use Ecto.Schema
  @derive {Saxy.Builder, name: "widget", attributes: [:type], children: [:content]}

  embedded_schema do
    field :type, :string
    field :content, :string
  end
end
```

What do you think?

**Comment by qcam** [comment]
Closed via #87

**Comment by jcelliott** [comment]
This looks great, thanks for adding that!
